service/storage_proxy.cc

```diff
@@ -159,6 +165,85 @@ class storage_proxy::remote {
+            return end_token(_range);
+        }
+    };
+
+    class write_request_info: public request_info {
+        remote& _remote;
+        table_schema_version _schema_version;
+        netw::msg_addr _msg_from;
+        partition_key_view _partition_key;
+    public:
+        write_request_info(remote& remote, table_schema_version schema_version, netw::msg_addr msg_from,
+            partition_key_view _partition_key)
+            : _remote(remote)
+            , _schema_version(schema_version)
+            , _msg_from(msg_from)
+            , _partition_key(std::move(_partition_key))
+        {
+        }
+        virtual future<schema_ptr> get_schema() const override {
+            return _remote.get_schema_for_write(_schema_version, _msg_from);
+        }
+        virtual dht::token get_token(const schema& schema) const override {
+            return schema.get_partitioner().get_token(schema, _partition_key);
+        }
+    };
+
+    future<std::optional<replica::stale_topology_exception>> try_apply_fence(fencing_token fence,
+        lowres_clock::time_point timeout,
+        const request_info& request_info)
+    {
+        while (true) {
+            const auto current_version = _sp._shared_token_metadata.get()->get_topology_version();
+            if (fence.topology_version == current_version) {
+                co_return std::nullopt;
+            }
+            if (fence.topology_version < current_version) {
+                auto schema = co_await request_info.get_schema();
+                const auto& ks_name = schema->ks_name();
+                const auto& ks = _sp._db.local().find_keyspace(ks_name);
+                auto erm = ks.get_effective_replication_map();
+                const auto token = request_info.get_token(*schema);
+                const auto& natural_endpoints = erm->get_natural_endpoints_without_node_being_replaced(token);
+                const auto& pending_endpoints = erm->get_token_metadata_ptr()->pending_endpoints_for(token, ks_name);
+
+                // Check if this replica is on the list of target replicas for mutation RPC.
+                // For read RPCs, if we are reading from the replica where the mutations go, it should be fine.
+                const auto token_owned =
+                    boost::find(natural_endpoints, utils::fb_utilities::get_broadcast_address()) != natural_endpoints.end() ||
+                    boost::find(pending_endpoints, utils::fb_utilities::get_broadcast_address()) != pending_endpoints.end();
+                if (token_owned) {
+                    co_return std::nullopt;
+                }
+                slogger.trace("Stale topology fenced, caller version {}, callee version {}",
+                    fence.topology_version, current_version);
+                co_return replica::stale_topology_exception(fence.topology_version, current_version);
+            }
+            co_await _sp._shared_token_metadata.wait_for_new_token_metadata(timeout);
```

Comment:
Alternative: why won't the caller send all the compatible topology versions? Then there's less expensive calculation at the time we transition between topologies.

Reply:
I think we could limit it to 1-or-2 versions.

Reply:
The idea is: two consecutive versions must be compatible, since there could be requests in-flight.

Let's say the current one is N. Before moving to N+1, ask all coordinators and all replicas to wait until N-1 is no longer in use. This is easy, just install a hook in that topology's destructor that records N-1 as no longer being used. If a node is marked as down, we can just skip it, and if it tries to use N-1 it will get fenced out.

Reply:
So the fencing token might be: uint64_t version, bool allow_previous.

Reply:
That is something we discussed with @gusev-p as well. The pros of having two compatible versions is less requests to be rejected because of incomparable versions the cons is more round trips between the topology coordinator  and other nodes to switch those versions. That does not solve the problem of getting a newer version though. There is no wait to atomically switch a cluster to another version without stopping the traffic.

Reply:
We don't need to atomically switch, since the current and next version are compatible. We just need to make sure that we don't move to the (next+1) version until all reachable nodes have confirmed they moved to (next). Unreachable nodes will need to become reachable AND update their version before they can accept traffic, unlike the current situation, where they only need to become reachable.

Reply:
How do you move to (next+1) atomically on all nodes? Some nodes will move to (next+1) earlier and will send requests with (next+1)  version to nodes with next. What they suppose to do?

Reply:
That can only be done if the topology versions are compatible, no? At some point the next+1 will not be compatible with next.

Lest suppose we what to move from {A, B, C} to { B, C, D} and lets consider writes only. Suppose {A, B, C} topology has version 1. First we need to move to a topology where writes goes to all 4 replicas {A, B, C, D}. This will be v=2. I is fine for A to accept both v1 and v2. No when we make sure everyone is on v2 we can start streaming. But after we start streaming A can no longer accept writes, so we need to move to v3 before we stream. But A cannot accept v2 any longer.

Reply:
turned `fencing_token` into `uint64_t version, bool allow_previous` as @avi requested